base/utfconvert.jl

```diff
@@ -0,0 +1,444 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+# Functions to convert to different UTF encodings
+
+# Quickly copy and set trailing \0
+@inline function fast_utf_copy(T::Type{UInt16}, len, dat)
+    @inbounds return UTF16String(setindex!(copy!(Vector{T}(len), dat), 0, len))
```

Comment:
It can result in undefined behavior if the array access turns out to be out of bounds. That inbounds annotation may end up turning off bounds checking everywhere inside the `UTF16String` constructor call, which could end up doing unsafe things. Maybe the current code is safe, but some future refactoring might make it unsafe and we'd start getting intermittent hard-to-reproduce segfaults instead of BoundsErrors. Hence the strong desire to limit the number of places we use these performance annotations.